app/src/main/java/com/github/se/eventradar/viewmodel/EventsOverviewViewModel.kt

```diff
@@ -30,6 +36,118 @@ constructor(
     checkUserLoginStatus()
   }
 
+  fun onSearchQueryChanged(query: String) {
+    _uiState.value = _uiState.value.copy(searchQuery = query)
+  }
+
+  fun onSearchActiveChanged(isSearchActive: Boolean) {
+    _uiState.value = _uiState.value.copy(isSearchActive = isSearchActive)
+  }
+
+  fun onFilterDialogOpen(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFilterDialogOpen = !state.value.isFilterDialogOpen)
+  }
+
+  fun onRadiusQueryChanged(radius: String) {
+    _uiState.value = _uiState.value.copy(radiusQuery = radius)
+  }
+
+  fun onFreeSwitchChanged(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFreeSwitchOn = !state.value.isFreeSwitchOn)
+  }
+
+  // TO DO: Function used to convert category selection to MVVM
+  //  fun onCategorySelectionChanged(category: EventCategory, isChecked: Boolean) {
+  //    val categoriesCheckedList = _uiState.value.categoriesCheckedList
+  //    Log.d("CategorySelection", "categoriesCheckedList $categoriesCheckedList")
+  //
+  //    if (isChecked) {
+  //      categoriesCheckedList.remove(category)
+  //    } else {
+  //      categoriesCheckedList.add(category)
+  //    }
+  //    _uiState.value = _uiState.value.copy(categoriesCheckedList = categoriesCheckedList)
+  //    Log.d("CategorySelection", "Category vm ${_uiState.value.categoriesCheckedList}")
+  //  }
+
+  fun onFilterApply(state: MutableStateFlow<EventsOverviewUiState> = _uiState) {
+    state.value = state.value.copy(isFilterActive = true)
+    filterEvents()
+  }
+
+  fun filterEvents() {
+    // Filter based on search query
+    val query = _uiState.value.searchQuery
+    val filteredEventsSearch =
+        _uiState.value.eventList.allEvents.filter {
+          it.eventName.contains(query, ignoreCase = true)
+        }
+
+    // Filter based on radius query
+    val radiusQuery = _uiState.value.radiusQuery
+
+    // For now, use a fixed user location - but this should be updated to use the user's actual
+    // location
+    val userLocation = Location(latitude = 38.92, longitude = 78.78, address = "Ecublens")
+    val filteredEventsRadius =
+        if (radiusQuery == "") {
+          _uiState.value.eventList.allEvents
+        } else {
+          _uiState.value.eventList.allEvents.filter { event ->
+            val distance = calculateDistance(userLocation, event.location)
+            distance <= radiusQuery.toDouble()
+          }
+        }
+
+    // Filter based on free switch
+    val isFreeSwitchOn = _uiState.value.isFreeSwitchOn
+    val filteredEventsFree =
+        _uiState.value.eventList.allEvents.filter {
+          if (isFreeSwitchOn) {
+            it.ticket.price == 0.0
+          } else {
+            true
+          }
+        }
+
+    // Filter based on categories selected
+    val categoriesCheckedList = _uiState.value.categoriesCheckedList
+    val filteredEventsCategory =
```

Comment:
imp: I think all the categories should be set to false at first, and in that case, no filter is applied so all the categories are shown (hopefully this helps with bug 1 of filter)

Reply:
Ok, I've set all categories to false first.